R/utils.R

```diff
@@ -166,3 +166,10 @@ rss = function() {  #5515 #5517
   # nocov end
 }
 
+# convert char to factor retaining order #4837
+fctr = function(x, levels=unique(x), ..., rev=FALSE) {
```

Comment:
We could push performance a bit by replacing `base::unique` with `funique`.
```r
funique = function(x, copy=FALSE) {
  o = forderv(x, sort=TRUE, retGrp=TRUE)
  if (attr(o, "maxgrpn", TRUE) <= 1L) { ## all unique
    if (copy) x = copy(x)
    return(x)
  }
  f = attr(o, "starts", exact=TRUE)
  if (length(o)) f = o[f]
  if (length(o <- forderv(f))) f = f[o] ## this rolls back to original order
  x[f]
}
```
but that adds a bit more code to maintain, and `fctr` is not kind of function to be commonly used in time critical paths. You call it once. Base `unique` is already fast, on 1e8 values we can push from 12s -> 6s, which is nice, but as said, don't want to introduce extra code to maintain but rather a convenience function on top of base R.